collector/fibrechannel_linux.go

```diff
@@ -0,0 +1,149 @@
+	"github.com/prometheus/procfs/sysfs"
+)
+
+const maxUint64 = ^uint64(0)
+
+type fibrechannelCollector struct {
+	fs          sysfs.FS
+	metricDescs map[string]*prometheus.Desc
+	logger      log.Logger
+	subsystem   string
+}
+
+func init() {
+	registerCollector("fibrechannel", defaultEnabled, NewFibreChannelCollector)
+}
+
+// NewFibreChannelCollector returns a new Collector exposing FibreChannel stats.
+func NewFibreChannelCollector(logger log.Logger) (Collector, error) {
+	var i fibrechannelCollector
+	var err error
+
+	i.fs, err = sysfs.NewFS(*sysPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open sysfs: %w", err)
+	}
+	i.logger = logger
+
+	// Detailed description for all metrics.
+	descriptions := map[string]string{
+		"dumped_frames_total":            "Number of dumped frames",
+		"loss_of_signal_total":           "Number of times signal has been lost",
+		"loss_of_sync_total":             "Number of failures on either bit or transmission word boundaries",
+		"rx_frames_total":                "Number of frames received",
+		"error_frames_total":             "Number of errors in frames",
+		"invalid_tx_words_total":         "Number of invalid words transmitted by host port",
+		"seconds_since_last_reset_total": "Number of seconds since last host port reset",
+		"tx_words_total":                 "Number of words transmitted by host port",
+		"invalid_crc_total":              "Invalid Cyclic Redundancy Check count",
+		"nos_total":                      "Number Not_Operational Primitive Sequence received by host port",
+		"fcp_packet_aborts_total":        "Number of aborted packets",
+		"rx_words_total":                 "Number of words received by host port",
+		"tx_frames_total":                "Number of frames transmitted by host port",
+		"link_failure_total":             "Number of times the host port link has failed",
+		"name":                           "Name of Fibre Channel HBA",
+		"speed":                          "Current operating speed",
+		"port_state":                     "Current port state",
+		"port_type":                      "Port type, what the port is connected to",
+		"symbolic_name":                  "Symoblic Name",
+		"node_name":                      "Node Name as hexadecimal string",
+		"port_id":                        "Port ID as string",
+		"port_name":                      "Port Name as hexadecimal string",
+		"fabric_name":                    "Fabric Name; 0 if PTP",
+		"dev_loss_tmo":                   "Device Loss Timeout in seconds",
+		"supported_classes":              "The FC classes supported",
+		"supported_speeds":               "The FC speeds supported",
+	}
+
+	i.metricDescs = make(map[string]*prometheus.Desc)
+	i.subsystem = "fibrechannel"
+
+	for metricName, description := range descriptions {
+		i.metricDescs[metricName] = prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, i.subsystem, metricName),
+			description,
+			[]string{"host"},
```

Comment:
There's a bit of a high probability that this will collide with service discovery label. Maybe we could use `fc_host` instead?

Reply:
done.